move to github actions

Datseris opened the pull request.

Datseris commented:
What the hell is going on with the tests!?!?!

Datseris commented:
Okay if @JonasIsensee , @tamasgal , @sebastianpech you guys have any idea why tests fail, please do help out here... Tests pass locally for me. :(

tamasgal commented:
Hmm... if I run `julia --project=. test/runtests.jl` I also get some errors:

Here is the full output (`juliat` is an alias to the line above)


```
░ tamasgal@greybox.local:DrWatson.jl  github-actions took 31s
░ 17:25:32 > juliat
┌ Warning:     Path separators in `savename` prefixes may break reproducibility on other OS.
│     The recommended way is using the `*dir()` methods or `joinpath` with
│     `savename` (e.g. `datadir("path", "to", "folder", savename("prefix", data))`).
└ @ DrWatson ~/Dev/DrWatson.jl/src/naming.jl:92
┌ Warning:     Path separators in `savename` prefixes may break reproducibility on other OS.
│     The recommended way is using the `*dir()` methods or `joinpath` with
│     `savename` (e.g. `datadir("path", "to", "folder", savename("prefix", data))`).
└ @ DrWatson ~/Dev/DrWatson.jl/src/naming.jl:92
┌ Warning:     Path separators in `savename` prefixes may break reproducibility on other OS.
│     The recommended way is using the `*dir()` methods or `joinpath` with
│     `savename` (e.g. `datadir("path", "to", "folder", savename("prefix", data))`).
└ @ DrWatson ~/Dev/DrWatson.jl/src/naming.jl:92
 Activating environment at `~/Dev/DrWatson.jl/Project.toml`
Base.active_project() = "/Users/tamasgal/Dev/DrWatson.jl/Project.toml"
Base.load_path_expand("@v$(VERSION.major).$(VERSION.minor)") = "/Users/tamasgal/.julia/environments/v1.5/Project.toml"
 Activating new environment at `~/test project/Project.toml`
   Updating registry at `~/.julia/registries/General`
   Updating git-repo `https://github.com/JuliaRegistries/General.git`
  Resolving package versions...
Updating `~/test project/Project.toml`
  [634d3b9d] + DrWatson v1.16.6
Updating `~/test project/Manifest.toml`
  [634d3b9d] + DrWatson v1.16.6
  [5789e2e9] + FileIO v1.4.5
  [ae029012] + Requires v1.1.2
  [3a884ed6] + UnPack v1.0.2
  [2a0f44e3] + Base64
  [ade2ca70] + Dates
  [b77e0a4c] + InteractiveUtils
  [76f85450] + LibGit2
  [8f399da3] + Libdl
  [56ddb016] + Logging
  [d6f4376e] + Markdown
  [44cfe95a] + Pkg
  [de0858da] + Printf
  [3fa0cd96] + REPL
  [9a3f8284] + Random
  [ea8e919c] + SHA
  [9e88b42a] + Serialization
  [6462fe0b] + Sockets
  [cf7118a7] + UUIDs
  [4ec0a83e] + Unicode
 Activating new environment at `~/test project/Project.toml`
  Resolving package versions...
Updating `~/test project/Project.toml`
  [634d3b9d] + DrWatson v1.16.6
Updating `~/test project/Manifest.toml`
  [634d3b9d] + DrWatson v1.16.6
  [5789e2e9] + FileIO v1.4.5
  [ae029012] + Requires v1.1.2
  [3a884ed6] + UnPack v1.0.2
  [2a0f44e3] + Base64
  [ade2ca70] + Dates
  [b77e0a4c] + InteractiveUtils
  [76f85450] + LibGit2
  [8f399da3] + Libdl
  [56ddb016] + Logging
  [d6f4376e] + Markdown
  [44cfe95a] + Pkg
  [de0858da] + Printf
  [3fa0cd96] + REPL
  [9a3f8284] + Random
  [ea8e919c] + SHA
  [9e88b42a] + Serialization
  [6462fe0b] + Sockets
  [cf7118a7] + UUIDs
  [4ec0a83e] + Unicode
 Activating new environment at `~/test project/Project.toml`
  Resolving package versions...
Updating `~/test project/Project.toml`
  [634d3b9d] + DrWatson v1.16.6
Updating `~/test project/Manifest.toml`
  [634d3b9d] + DrWatson v1.16.6
  [5789e2e9] + FileIO v1.4.5
  [ae029012] + Requires v1.1.2
  [3a884ed6] + UnPack v1.0.2
  [2a0f44e3] + Base64
  [ade2ca70] + Dates
  [b77e0a4c] + InteractiveUtils
  [76f85450] + LibGit2
  [8f399da3] + Libdl
  [56ddb016] + Logging
  [d6f4376e] + Markdown
  [44cfe95a] + Pkg
  [de0858da] + Printf
  [3fa0cd96] + REPL
  [9a3f8284] + Random
  [ea8e919c] + SHA
  [9e88b42a] + Serialization
  [6462fe0b] + Sockets
  [cf7118a7] + UUIDs
  [4ec0a83e] + Unicode
Saving tools: Error During Test at /Users/tamasgal/Dev/DrWatson.jl/test/runtests.jl:7
  Got exception outside of a @test
  LoadError: ArgumentError: Package DrWatson not found in current path:
  - Run `import Pkg; Pkg.add("DrWatson")` to install the DrWatson package.

  Stacktrace:
   [1] require(::Module, ::Symbol) at ./loading.jl:893
   [2] include(::String) at ./client.jl:457
   [3] top-level scope at /Users/tamasgal/Dev/DrWatson.jl/test/runtests.jl:7
   [4] top-level scope at /Users/julia/buildbot/worker/package_macos64/build/usr/share/julia/stdlib/v1.5/Test/src/Test.jl:1115
   [5] top-level scope at /Users/tamasgal/Dev/DrWatson.jl/test/runtests.jl:7
   [6] top-level scope at /Users/julia/buildbot/worker/package_macos64/build/usr/share/julia/stdlib/v1.5/Test/src/Test.jl:1115
   [7] top-level scope at /Users/tamasgal/Dev/DrWatson.jl/test/runtests.jl:4
   [8] include(::Function, ::Module, ::String) at ./Base.jl:380
   [9] include(::Module, ::String) at ./Base.jl:368
   [10] exec_options(::Base.JLOptions) at ./client.jl:296
   [11] _start() at ./client.jl:506
  in expression starting at /Users/tamasgal/Dev/DrWatson.jl/test/stools_tests.jl:1

Produce or Save: Error During Test at /Users/tamasgal/Dev/DrWatson.jl/test/runtests.jl:8
  Got exception outside of a @test
  LoadError: ArgumentError: Package DrWatson not found in current path:
  - Run `import Pkg; Pkg.add("DrWatson")` to install the DrWatson package.

  Stacktrace:
   [1] require(::Module, ::Symbol) at ./loading.jl:893
   [2] include(::String) at ./client.jl:457
   [3] top-level scope at /Users/tamasgal/Dev/DrWatson.jl/test/runtests.jl:8
   [4] top-level scope at /Users/julia/buildbot/worker/package_macos64/build/usr/share/julia/stdlib/v1.5/Test/src/Test.jl:1115
   [5] top-level scope at /Users/tamasgal/Dev/DrWatson.jl/test/runtests.jl:8
   [6] top-level scope at /Users/julia/buildbot/worker/package_macos64/build/usr/share/julia/stdlib/v1.5/Test/src/Test.jl:1115
   [7] top-level scope at /Users/tamasgal/Dev/DrWatson.jl/test/runtests.jl:4
   [8] include(::Function, ::Module, ::String) at ./Base.jl:380
   [9] include(::Module, ::String) at ./Base.jl:368
   [10] exec_options(::Base.JLOptions) at ./client.jl:296
   [11] _start() at ./client.jl:506
  in expression starting at /Users/tamasgal/Dev/DrWatson.jl/test/savefiles_tests.jl:1

Collect Results: Error During Test at /Users/tamasgal/Dev/DrWatson.jl/test/runtests.jl:9
  Got exception outside of a @test
  LoadError: ArgumentError: Package DrWatson not found in current path:
  - Run `import Pkg; Pkg.add("DrWatson")` to install the DrWatson package.

  Stacktrace:
   [1] require(::Module, ::Symbol) at ./loading.jl:893
   [2] include(::String) at ./client.jl:457
   [3] top-level scope at /Users/tamasgal/Dev/DrWatson.jl/test/runtests.jl:9
   [4] top-level scope at /Users/julia/buildbot/worker/package_macos64/build/usr/share/julia/stdlib/v1.5/Test/src/Test.jl:1115
   [5] top-level scope at /Users/tamasgal/Dev/DrWatson.jl/test/runtests.jl:9
   [6] top-level scope at /Users/julia/buildbot/worker/package_macos64/build/usr/share/julia/stdlib/v1.5/Test/src/Test.jl:1115
   [7] top-level scope at /Users/tamasgal/Dev/DrWatson.jl/test/runtests.jl:4
   [8] include(::Function, ::Module, ::String) at ./Base.jl:380
   [9] include(::Module, ::String) at ./Base.jl:368
   [10] exec_options(::Base.JLOptions) at ./client.jl:296
   [11] _start() at ./client.jl:506
  in expression starting at /Users/tamasgal/Dev/DrWatson.jl/test/update_results_tests.jl:1

Parameter Customization: Error During Test at /Users/tamasgal/Dev/DrWatson.jl/test/runtests.jl:10
  Got exception outside of a @test
  LoadError: ArgumentError: Package DrWatson not found in current path:
  - Run `import Pkg; Pkg.add("DrWatson")` to install the DrWatson package.

  Stacktrace:
   [1] require(::Module, ::Symbol) at ./loading.jl:893
   [2] include(::String) at ./client.jl:457
   [3] top-level scope at /Users/tamasgal/Dev/DrWatson.jl/test/runtests.jl:10
   [4] top-level scope at /Users/julia/buildbot/worker/package_macos64/build/usr/share/julia/stdlib/v1.5/Test/src/Test.jl:1115
   [5] top-level scope at /Users/tamasgal/Dev/DrWatson.jl/test/runtests.jl:10
   [6] top-level scope at /Users/julia/buildbot/worker/package_macos64/build/usr/share/julia/stdlib/v1.5/Test/src/Test.jl:1115
   [7] top-level scope at /Users/tamasgal/Dev/DrWatson.jl/test/runtests.jl:4
   [8] include(::Function, ::Module, ::String) at ./Base.jl:380
   [9] include(::Module, ::String) at ./Base.jl:368
   [10] exec_options(::Base.JLOptions) at ./client.jl:296
   [11] _start() at ./client.jl:506
  in expression starting at /Users/tamasgal/Dev/DrWatson.jl/test/customize_savename.jl:1

Test Summary:             | Pass  Error  Total
DrWatson                  |  129      4    133
  Naming                  |   48            48
  Parse savename          |   18            18
  Project Setup           |   63            63
  Saving tools            |           1      1
  Produce or Save         |           1      1
  Collect Results         |           1      1
  Parameter Customization |           1      1
ERROR: LoadError: Some tests did not pass: 129 passed, 0 failed, 4 errored, 0 broken.
in expression starting at /Users/tamasgal/Dev/DrWatson.jl/test/runtests.jl:3
```

Datseris commented:
I don't understand. What has broken so far?

tamasgal commented:
This is one broken tests, related to the git version string:

```
░ tamasgal@greybox.local:DrWatson.jl  github-actions took 8s
░ 17:35:21 > juliap test/stools_tests.jl
Test Failed at /Users/tamasgal/Dev/DrWatson.jl/test/stools_tests.jl:34
  Expression: d == d_new
   Evaluated: Dict(:y => 4,:x => 3) == Dict{Symbol,Any}(:y => 4,:gitcommit => "v1.16.6-8-g3078b31",:x => 3)
ERROR: LoadError: There was an error during testing
in expression starting at /Users/tamasgal/Dev/DrWatson.jl/test/stools_tests.jl:34
```

tamasgal commented:
The `savefiles_tests.jl` fail because `BSON` is in extras and is not picked up:

```
░ tamasgal@greybox.local:DrWatson.jl  github-actions took 5s
░ 17:35:29 1 > juliap test/savefiles_tests.jl
Error encountered while saving "N=50_T=1000_every=10_seed=1111_Δt=0.05.bson".

Fatal error:
ERROR: LoadError: ArgumentError: Package BSON not found in current path:
- Run `import Pkg; Pkg.add("BSON")` to install the BSON package.
```

tamasgal commented:
I am confused about those extras-errors. Also the `customize_savename.jl` fails:

```
░ tamasgal@greybox.local:DrWatson.jl  github-actions
░ 17:38:47 130 > juliap test/customize_savename.jl
ERROR: LoadError: ArgumentError: Package Parameters not found in current path:
- Run `import Pkg; Pkg.add("Parameters")` to install the Parameters package.
```

when I run the tests from the REPL (`]test`), everything works correctly.

tamasgal commented:
Since those extras-targets are used (I have not used them much so far), I think that the invocation from the CI is somehow broken, but I am not sure...

tamasgal commented:
How do you launch the tests on your machine? Also via REPL and then `]test`?

I also checked the diff of `ci.yml` to my other projects which work but I don't see any obvious hint...

Since `julia --project=. test/runtests.jl` is not working here (due to the extras deps), I also tried running the command which afaik is also used in CI, and that fails with many errors:

```
julia -e 'using Pkg; Pkg.add(PackageSpec(path=pwd())); Pkg.build("DrWatson"); Pkg.test("DrWatson"; coverage = true)'
```

Here are some errors:

```
Saving tools: Test Failed at /Users/tamasgal/.julia/packages/DrWatson/clwkq/test/stools_tests.jl:5
  Expression: com isa String
   Evaluated: nothing isa String

Saving tools: Test Failed at /Users/tamasgal/.julia/packages/DrWatson/clwkq/test/stools_tests.jl:8
  Expression: com !== nothing
   Evaluated: nothing !== nothing

Saving tools: Test Failed at /Users/tamasgal/.julia/packages/DrWatson/clwkq/test/stools_tests.jl:9
  Expression: typeof(com) <: String

Saving tools: Error During Test at /Users/tamasgal/.julia/packages/DrWatson/clwkq/test/stools_tests.jl:10
  Test threw exception
  Expression: com[1] == 'v'
  MethodError: no method matching getindex(::Nothing, ::Int64)

Saving tools: Test Failed at /Users/tamasgal/.julia/packages/DrWatson/clwkq/test/stools_tests.jl:17
  Expression: haskey(d, (keytype(d))(:gitcommit))

Saving tools: Error During Test at /Users/tamasgal/.julia/packages/DrWatson/clwkq/test/stools_tests.jl:18
  Test threw exception
  Expression: d[(keytype(d))(:gitcommit)] |> typeof <: String
  KeyError: key :gitcommit not found

Saving tools: Test Failed at /Users/tamasgal/.julia/packages/DrWatson/clwkq/test/stools_tests.jl:17
  Expression: haskey(d, (keytype(d))(:gitcommit))
```

and the summary:

```
Test Summary:             | Pass  Fail  Error  Total
DrWatson                  |  284    13     13    310
  Naming                  |   48                  48
  Parse savename          |   18                  18
  Project Setup           |   63                  63
  Saving tools            |   66     9      9     84
  Produce or Save         |   43     4      4     51
  Collect Results         |   23                  23
  Parameter Customization |   23                  23
ERROR: LoadError: Some tests did not pass: 284 passed, 13 failed, 13 errored, 0 broken.
```

Datseris commented:
> How do you launch the tests on your machine? Also via REPL and then ]test?

Yeap, exactly. and I see no errors with this approach

tamasgal commented:
OK, the REPL approach would be `julia --project=. -e 'using Pkg; Pkg.test("DrWatson"; coverage = true)'`

This works for me either... but the CI does something different. Maybe it clashes with the activation of other projects while running the test-script? 😕

tamasgal commented:
This is the command which is executed, as far as I can see:

https://github.com/julia-actions/julia-runtest/blob/b32c73084a913834335f8ff0ce6f6cad610a6fcf/action.yml#L48

Which is then expanded to (correct me if I am wrong):

```
julia --color=yes --check-bounds=yes --inline=yes --depwarn=yes --project=. -e 'using Pkg; Pkg.test(coverage=true)'
```

But that also works for me, locally... 😡

JonasIsensee commented:
```
Project Setup: Error During Test at /home/runner/work/DrWatson.jl/DrWatson.jl/test/runtests.jl:6
  Got exception outside of a @test
  LoadError: GitError(Code:ENOTFOUND, Class:Config, config value 'user.name' was not found)
  Stacktrace:
   [1] macro expansion at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.5/LibGit2/src/error.jl:106 [inlined]
   [2] default_signature(::LibGit2.GitRepo) at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.5/LibGit2/src/signature.jl:69
   [3] LibGit2.Signature(::LibGit2.GitRepo) at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.5/LibGit2/src/signature.jl:25
   [4] commit at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.5/LibGit2/src/commit.jl:113 [inlined]
   [5] initialize_project(::String, ::String; force::Bool, readme::Bool, authors::Nothing, git::Bool, placeholder::Bool) at /home/runner/work/DrWatson.jl/DrWatson.jl/src/project_setup.jl:280
```

Datseris commented:
I've added the lines 
```
          - git config --global user.name "GitUser"
          - git config --global user.email "git@us.er"
```
to the CI script, but maybe it diidn't help. I added the LibGit command now.

tamasgal commented:
This might be a bit desperate, but maybe it's worth trying to use environment variables in case the `git config` somehow fails (and also the `LibGit2.default_signature`)... as said, it's desperate but who knows.

Like this in `ci.yml`:

```
       - GIT_AUTHOR_NAME=GitUser GIT_AUTHOR_EMAIL=git@us.er julia --project=docs -e '
           using Pkg
           Pkg.develop(PackageSpec(path=pwd()))
           Pkg.instantiate()'
```

tamasgal commented:
At least it's now easily reproducible (e.g. via Docker), thanks to @JonasIsensee who spotted the right line. I always get confused in these monochromatic log outputs. GitLab CI log formatting is way nicer 😉 

```
░ tgal@staticbox:~ took 28s
░ 23:52:57 > docker run -it julia:1.5.3
               _
   _       _ _(_)_     |  Documentation: https://docs.julialang.org
  (_)     | (_) (_)    |
   _ _   _| |_  __ _   |  Type "?" for help, "]?" for Pkg help.
  | | | | | | |/ _` |  |
  | | |_| | | | (_| |  |  Version 1.5.3 (2020-11-09)
 _/ |\__'_|_|_|\__'_|  |  Official https://julialang.org/ release
|__/                   |

julia> using LibGit2

julia> repo = LibGit2.init("narf")
LibGit2.GitRepo("/narf")

julia> LibGit2.commit(repo, "Initial commit")
ERROR: GitError(Code:ENOTFOUND, Class:Config, config value 'user.name' was not found)
Stacktrace:
 [1] macro expansion at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.5/LibGit2/src/error.jl:106 [inlined]
 [2] default_signature(::GitRepo) at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.5/LibGit2/src/signature.jl:69
 [3] LibGit2.Signature(::GitRepo) at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.5/LibGit2/src/signature.jl:25
 [4] commit(::GitRepo, ::String) at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.5/LibGit2/src/commit.jl:113
 [5] top-level scope at REPL[3]:1
```

tamasgal commented:
Btw. I also tried installed `git` in the Julia docker (`apt update && apt install git`) and set the environment variable and was able to use `git` in the command-line, however inside a `Julia` session I get the same error, so `LibGit2` ignores those, as it seems.

However, using `git config --global ...` works:

```
root@af1631535f7e:/# git config --global user.name "GitUser"
root@af1631535f7e:/# git config --global user.email "git@us.er"
root@af1631535f7e:/# julia
               _
   _       _ _(_)_     |  Documentation: https://docs.julialang.org
  (_)     | (_) (_)    |
   _ _   _| |_  __ _   |  Type "?" for help, "]?" for Pkg help.
  | | | | | | |/ _` |  |
  | | |_| | | | (_| |  |  Version 1.5.3 (2020-11-09)
 _/ |\__'_|_|_|\__'_|  |  Official https://julialang.org/ release
|__/                   |

julia> using LibGit2

julia> repo = LibGit2.init("narf")
LibGit2.GitRepo("/narf")

julia> LibGit2.commit(repo, "Initial commit")
GitHash("775e58dd7db754ede0a08ac020bd3234015c1487")
```

tamasgal commented on test/stools_tests.jl:
```diff
@@ -7,7 +7,9 @@ com = gitdescribe(@__DIR__)
 @test typeof(com) <: String
-@test com[1] == 'v' # test that it has a version tag
+# TODO: why?
+# @test com[1] == 'v' # test that it has a version tag
```
This one fails since `com` is just a commit hash since there is no tag visible to `gitdescribe`. Maybe `@__DIR__` is not pointing to DrWatsons directory or for whatever reason...

I think we can ditch that test completely. It should be enough that it's not `nothing`, at least from the reproducibility perspective.

tamasgal commented:
Looks good now. I need to get sleep, it was a long Sunday ;)

tamasgal commented:
@Datseris just a tiny ping, I guess your personal spam filter was activated on this PR `;D` I think this should be OK now. The coverage went down a bit though `;)`

Datseris commented:
Yo Tamas, sorry, I actually missed the last message here. Thought this wasn't fixed. 

But also feel free to merge things on your own accord, you're also a main dev, no reason to wait for me if you're confident for merges
